crypto/schnorr.c

```diff
@@ -0,0 +1,252 @@
+
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+
+int jacobi(const bignum256 *_n, const bignum256 *_k) {
+  int t = 0;
+  int k_is_one = 0;
+  uint32_t r = 0;
+  bignum256 n = {0}, k = {0}, temp = {0};
+
+  assert(!bn_is_zero(_k) && bn_is_odd(_k));
+
+  bn_copy(_n, &n);
+  bn_copy(_k, &k);
+
+  bn_mod(&n, &k);
+
+  while (!bn_is_zero(&n)) {
+    while (bn_is_even(&n)) {
+      bn_rshift(&n);
+      // jacobi(2 * n, k) = jacobi(n, k) if k = 1 (mod 8) or k = 7 (mod 8)
+      // jacobi(2 * n, k) = -jacobi(n, k) if k = 3 (mod 8) or k = 5 (mod 8)
+      r = k.val[0] & 0x07;
+      t ^= (r == 3 || r == 5);
+    }
+
+    bn_copy(&n, &temp);
+    bn_copy(&k, &n);
+    bn_copy(&temp, &k);
+
+    // jacobi(n, k) = jacobi(k, n) if k = n = 1 (mod 4)
+    // jacobi(n, k) = -jacobi(k, n) if k = n = 3 (mod 4)
+    t ^= ((n.val[0] & k.val[0] & 3) == 3);
+
+    bn_mod(&n, &k);
+  }
+
+  k_is_one = bn_is_one(&k);
+
+  // Cleanup
+  memzero(&n, sizeof(n));
+  memzero(&k, sizeof(k));
+  memzero(&temp, sizeof(temp));
+
+  // Map t: [0] => 1, [1] => -1
+  t = -2 * t + 1;
+
+  return k_is_one * t;
+}
+
+int is_non_quad_residue(const bignum256 *n, const bignum256 *prime) {
+  return jacobi(n, prime) == -1;
+}
+
+// Compute sha256(sha256(msg))
+void double_sha256(const uint8_t *msg, uint32_t msg_len,
+                   uint8_t digest[SHA256_DIGEST_LENGTH]) {
+  SHA256_CTX ctx = {0};
+
+  sha256_Init(&ctx);
+  sha256_Update(&ctx, msg, msg_len);
+  sha256_Final(&ctx, digest);
+  sha256_Init(&ctx);
+  sha256_Update(&ctx, digest, SHA256_DIGEST_LENGTH);
+  sha256_Final(&ctx, digest);
+}
+
+/*
+ * Init the HMAC with additional data specific to Schnorr. This prevents from
+ * leaking the private key in the case the same message is signed with both
+ * Schnorr and ECDSA.
+ */
+void init_rfc6979_schnorr(const uint8_t *priv_key, const uint8_t *hash,
+                          rfc6979_state *state) {
+  uint8_t hmac_data[SHA256_DIGEST_LENGTH + 16] = {0};
+
+  memcpy(hmac_data, hash, SHA256_DIGEST_LENGTH);
+  memcpy(hmac_data + SHA256_DIGEST_LENGTH, "Schnorr+SHA256  ", 16);
+
+  hmac_drbg_init(state, priv_key, 32, hmac_data, SHA256_DIGEST_LENGTH + 16);
+}
+
+int generate_k_schnorr(const ecdsa_curve *curve, bignum256 *k,
+                       rfc6979_state *state) {
+  int i = 0;
+
+  for (i = 0; i < 10000; i++) {
+    generate_k_rfc6979(k, state);
+    // If k is too big or too small, we don't like it
+    if (bn_is_zero(k) || !bn_is_less(k, &curve->order)) {
+      continue;
+    }
+
+    return 0;
+  }
+
+  return 1;
+}
+
+// e = H(Rx, pub_key, msg_hash)
+static void calc_e(const ecdsa_curve *curve, const bignum256 *Rx,
+                   const uint8_t pub_key[33], const uint8_t *msg_hash,
+                   bignum256 *e) {
+  uint8_t Rxbuf[32] = {0};
+  SHA256_CTX ctx = {0};
+  uint8_t digest[SHA256_DIGEST_LENGTH] = {0};
+
+  bn_write_be(Rx, Rxbuf);
+
+  sha256_Init(&ctx);
+  sha256_Update(&ctx, Rxbuf, sizeof(Rxbuf));
+  sha256_Update(&ctx, pub_key, 33);
```

Comment:
`sizeof(pub_key)` is better than `33`.

Reply:
I don't think this can work since `pub_key` is a parameter, this would be `sizeof(uint8_t *)`

Reply:
Absolutely right.